### DISCUSSION TITLE
Flexible starting components for modules.

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#765] [Ideas] [answered]
When we add a module to a page, the index.razor is the default starting module.  This makes perfect sense.  In the case of lots of little one-page modules, when combined into a multi-module project, you end up with lots of components called index.razor separated by folders and namespaces.  In this case it would be handy to have less namespaces and more differently named razor files.  

example current:
- module
  - SomeAdminModule
    - Index.razor
  - AnotherAdminModule
    - Index.razor

example proposed:
- module
  - SomeAdmin.razor
  - AnotherAdmin.Razor

This configuration has some usability advantages.  When most of your components are named index.razor, switching between editors can get confusing.  
![image](https://user-images.githubusercontent.com/311063/94643903-b9c85600-029c-11eb-818e-9f27334ae81b.png)

Options for implementation:
- Option one: startup component selector
- - To facilitate selecting one of these alternate components as the starting component for the module will require reworking code in the module loading, adjusting the PageModule schema to add a StartComponent column, and adjusting the Add Module UI to allow the selection of the starting component from a list of components that inherit from modulebase.
- Option two: decorator interface
- - An alternative solution would be to decorate components that impliment modulebase with an interface like IModuleIndex and then list each of these as modules.  This might be a better short-term solution, but I think eventually, we'll end up with thousands of "modules" when we really want a fewer modules with numerous entry-points each.

**Comment by sbwalker** [comment]
I think the simplest option and the one which may preserve the existing performance characteristics is to add an optional property to the ModuleDefinition model named DefaultComponent - which would allow a developer to specify the value in their IModule interface. This is the purpose of the IModule interface - to provide metadata which governs the behavior of the module.

**Reply by markdav-is** [reply]
Nice.  I'll look at that and see how it might apply to the many-module project we have rolling.

**Reply by PizzaConsole** [reply]
I like that idea

**Reply by PizzaConsole** [reply]
@sbwalker After thinking about it (correct me if I am wrong), I actually think that this is something that should be set on the Actions Property in ModuleBase. 
For example, you have a page named `StartingPage.razor` in a Custom Module you should be able to set the Actions to `public override string Actions { get { return "Index"; } }` and that should automatically determine that it is the starting page. I like keeping the term "Index" it is a universal term in Web development. 
Currently this works correctly IF you are already inside the Moduel e.g. `http://localhost:44357/private/*/27` or `http://localhost:44357/private/*/27/Index` would route to the Page whose Actions specifies "Index".
Thoughts?

**Reply by PizzaConsole** [reply]
I made a PR for this: #770

**Reply by sbwalker** [reply]
By default there is a 1:1 mapping between the action and module component filename ( ie. "edit" maps to Edit.razor ). So the purpose of the Actions property is to allow a developer to control the mapping and even allow multiple actions to map to a specific module component. For example, the Blog module uses this approach to indicate that the "add" and "edit" action both map to the Edit.razor component: ie. Actions = Add,Edit". And you are correct that it can also be used to override the default 1:1 mapping - ie. in the case where you have a really long descriptive component filename and want to map it to a short action - "reportA" = ReallyPowerfulReportVersionA.razor. 

So what you are suggesting would definitely work - mapping a component to "index". The question is what would be more intuitive to developers. There are pros and cons to each approach outlined.

Actions - allow developers to map a component to the "Index" action
- Pro: can be implemented directly in components and does not require an IModule implementation for a module
- Con: implicit as it still relies indirectly on the default "Index" action which is essentially a magic string, and magic strings always impose a learning curve for newer developers.  Also it would be easy during copy/paste for a developer to mistakenly include the Action property definition in multiple components within the same module which would lead to undesired behavior.

IModule - a new optional property for either DefaultAction or DefaultComponent which overrides the default "Index" action for that module
- Pro: explicitly declared with all of the other module metadata which control the overall behavior of a module.
- Con: requires implementation of IModule interface for module

**Reply by sbwalker** [reply]
The other option of course is to support both approaches - it makes the core implementation a bit more complicated but these solutions could complement each other

**Reply by PizzaConsole** [reply]
Both sounds intriguing. Honestly I feel like the PR i did is just default behavior that the framework should already have. Like I said the Actions behavior already works when inside the module. I would be in favour of doing both.

> Also it would be easy during copy/paste for a developer to mistakenly include the Action property definition in multiple components within the same module which would lead to undesired behavior.

Isn't this technically already a con of how the actions currently work?

**Reply by PizzaConsole** [reply]
@markdav-is This feature is now available via the latest PR merge

**Reply by sbwalker** [reply, marked as answer]
I am also introducing the DefaultAction option described above - so you will have a couple of solutions to your issue.

**Reply by markdav-is** [reply]
Wow. Thanks!  I really appreciate it.

**Comment by PizzaConsole** [comment]
I am very curious, how you are using the modules. What would be the reason for so many single page modules?

**Reply by markdav-is** [reply]
config settings.  pages and pages of config settings.
